### ISSUE TITLE
[Performance] Profiling on CUDA shows confusing values

### ISSUE THREAD

**Issue by Dr-Electron** [microsoft/onnxruntime#20398]
### Describe the issue

I tried profiling my UNET running on CUDA. And while looking at it in Perfetto, I saw that `SequentialExecutor::Execute` actually took only a fraction of the whole `model_run` time:
<img width="1574" alt="perfetto" src="https://github.com/microsoft/onnxruntime/assets/61637085/8c106490-de03-4aac-b2d6-d36d144eb31d">
So I started to wonder what is happening the rest of the time? I set up nsys and tried again:
<img width="1562" alt="nsys" src="https://github.com/microsoft/onnxruntime/assets/61637085/b842adad-1d32-42c6-9e88-4b93ad4a748f">
So far, so good. I don't yet understand why I see `run` two times, but the CUDA one shows roughly the same 200ms as the result from onnxruntime.
But the actual CUDA kernel execution time starts way later, as shown in the onnxruntime profiling. So I checked. In Perfetto it takes around 40ms until `SequentialExecutor::Execute` starts:
<img width="1573" alt="perfetto_40" src="https://github.com/microsoft/onnxruntime/assets/61637085/55e62c79-ebf8-4065-91ba-27a7fcbefb8e">
Which is roughly the time it takes to make the first kernel related `CUDA API` calls looking at nsys (Look at the CUDA API row. For this example, I zoomed out, but this is the section where the Kernel specific API calls start):
<img width="1559" alt="nsys_40" src="https://github.com/microsoft/onnxruntime/assets/61637085/401d522b-9496-4052-942b-2d534a8aaf81">
And afterward `SequentialExecutor::Execute` takes around 5ms:
<img width="1575" alt="perfetto_5" src="https://github.com/microsoft/onnxruntime/assets/61637085/c8c9159d-f944-4daa-8cee-6eaf203a9e08">
Which looking at nsys is the time it takes for all API calls to finish:
<img width="1563" alt="nsys_5" src="https://github.com/microsoft/onnxruntime/assets/61637085/4c157f3b-2dac-4a94-8633-17000db15e1b">

So maybe I understood it wrong, I never really worked with CUDA, only with OpenCL, but to me, it seems what onnxruntime is profiling here is not the actual execution time of the kernels, but the CUDA API times for them. Which seems quite weird to me, as the execution times are probably what the user is interested in. At least both times should be shown.
This could also be related to https://github.com/microsoft/onnxruntime/issues/19930

Thanks for looking at this and feel free to ask more question if I was unclear somewhere.

### To reproduce

1) Get the UNET I used as a test.
2) Execute the following Python code with nsys and compare results with onnxruntime profiling.

```python
import time

import numpy as np
import onnxruntime

import nvtx

providers = ['CUDAExecutionProvider']

session_options_profiling = onnxruntime.SessionOptions()
session_options_profiling.enable_profiling = True
session_options_profiling.log_severity_level = 4
session_options_profiling.log_verbosity_level = 0

session_profiling = onnxruntime.InferenceSession('./unet_13_1080p.onnx', sess_options=session_options_profiling, providers=providers)

# Get input and output names
input_name = session_profiling.get_inputs()[0].name
label_name = session_profiling.get_outputs()[0].name

# Profiling
# Warmup run
with nvtx.annotate("warmup"):
    session_profiling.run([label_name], {input_name: np.random.random((1, 3, 1080, 1920)).astype(np.float32)})
start = time.perf_counter()
with nvtx.annotate("run"):
            session_profiling.run([label_name],  {input_name: np.random.random((1, 3, 1080, 1920)).astype(np.float32)})
stop = time.perf_counter()
print(f'{(stop - start):.3f}')
session_profiling.end_profiling()
```

### Urgency

Not urgent

### Platform

Windows

### OS Version

Windows IoT LTSC 21H2 19044.4291

### ONNX Runtime Installation

Built from Source

### ONNX Runtime Version or Commit ID

1.18.0.dev20240316001

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

CUDA

### Execution Provider Library Version

CUDA 11.8

### Model File

https://github.com/cagery/unet-onnx/blob/d3d5670ed82cfa14e046cb5b377e38f7943746ac/unet_13_1080p.onnx

### Is this a quantized model?

No

**Comment by Dr-Electron** [comment]
Not stale

**Comment by tianleiwu** [comment]
To profile CUDA, you will need build onnxruntime with option `--cmake_extra_defines onnxruntime_ENABLE_NVTX_PROFILE=ON`. Example build command line in Windows
```
call "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" amd64 
build.bat --cmake_generator "Visual Studio 17 2022" --config Release --build_dir build\cuda12.6_nvtx --build_wheel ^
          --parallel 4 --nvcc_threads 1 --build_shared_lib ^
          --use_cuda --cuda_version "12.6" --cuda_home "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.6" ^
          --cudnn_home "C:\CuDNN\9.3.0.75_cuda12" ^
          --cmake_extra_defines "CMAKE_CUDA_ARCHITECTURES=61" ^
          --skip_tests --compile_no_warning_as_error ^
          --use_binskim_compliant_compile_flags ^
          --cmake_extra_defines onnxruntime_ENABLE_NVTX_PROFILE=ON --enable_cuda_line_info
pip install build\cuda12.6_nvtx\Release\Release\dist\*.whl
```

Example of UNet (multiple iterations) with IO binding (inputs and outputs are in GPU memory) is like the following:
![Image](https://github.com/user-attachments/assets/ca94e788-115b-4f18-baaf-412bbe7eeb16)

Zoom it and you can see the execution time of each onnx node:
![Image](https://github.com/user-attachments/assets/494607b1-f77b-4cc2-a7a4-d1276d523c71)

If I do not use IO binding, the profiling result looks confusing (as cudaMemcpyAsync occupy most of the time for unet):
![Image](https://github.com/user-attachments/assets/d59c5f62-1c9c-4927-b81c-c6fe16074cef)

In summary, try custom build and bind inputs/outputs to tensors in GPU.
